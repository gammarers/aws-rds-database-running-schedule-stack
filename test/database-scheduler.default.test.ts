import { App } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { DatabaseType, RdsDatabaseRunningScheduleStack } from '../src';

describe('RdsDatabaseRunningScheduler Default Testing', () => {

  describe('default schedule', () => {
    const app = new App();

    const stack = new RdsDatabaseRunningScheduleStack(app, 'RdsDatabaseRunningScheduleStack', {
      targets: [
        {
          type: DatabaseType.CLUSTER,
          identifiers: ['db-cluster-1a'],
          startSchedule: {
            timezone: 'UTC',
          },
          stopSchedule: {
            timezone: 'UTC',
          },
        },
        {
          type: DatabaseType.INSTANCE,
          identifiers: ['db-instance-1a'],
          startSchedule: {
            timezone: 'UTC',
          },
          stopSchedule: {
            timezone: 'UTC',
          },
        },
      ],
    });

    const template = Template.fromStack(stack);

    // Schedule execution role
    it('Should have schedule execution role', async () => {
      template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
        RoleName: Match.stringLikeRegexp('stop-db-schedule-.*-exec-role'),
        AssumeRolePolicyDocument: Match.objectEquals({
          Version: '2012-10-17',
          Statement: Match.arrayWith([
            Match.objectEquals({
              Effect: 'Allow',
              Principal: {
                Service: 'scheduler.amazonaws.com',
              },
              Action: 'sts:AssumeRole',
            }),
          ]),
        }),
        Policies: Match.arrayEquals([
          {
            PolicyName: 'rds-instance-stop-policy',
            PolicyDocument: Match.objectEquals({
              Version: '2012-10-17',
              Statement: [
                Match.objectEquals({
                  Effect: 'Allow',
                  Action: Match.arrayEquals([
                    'rds:StopDBInstance',
                    'rds:StartDBInstance',
                  ]),
                  Resource: {
                    'Fn::Join': [
                      '',
                      [
                        'arn:aws:rds:*:',
                        {
                          Ref: 'AWS::AccountId',
                        },
                        ':db:*',
                      ],
                    ],
                  },
                }),
              ],
            }),
          },
          {
            PolicyName: 'rds-cluster-stop-policy',
            PolicyDocument: Match.objectEquals({
              Version: '2012-10-17',
              Statement: [
                Match.objectEquals({
                  Effect: 'Allow',
                  Action: Match.arrayEquals([
                    'rds:StopDBCluster',
                    'rds:StartDBCluster',
                  ]),
                  Resource: {
                    'Fn::Join': [
                      '',
                      [
                        'arn:aws:rds:*:',
                        {
                          Ref: 'AWS::AccountId',
                        },
                        ':cluster:*',
                      ],
                    ],
                  },
                }),
              ],
            }),
          },
        ]),
      }));
    });

    // Start Cluster Schedule testing
    it('Should have Cluster Start Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-start-db-cluster-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'UTC',
        ScheduleExpression: 'cron(50 7 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:rds:startDBCluster',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('SchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: Match.stringLikeRegexp('{"DbClusterIdentifier":"db-cluster-1a"}'),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    // Stop Cluster Schedule testing
    it('Should have Cluster Stop Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-stop-db-cluster-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'UTC',
        ScheduleExpression: 'cron(10 21 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:rds:stopDBCluster',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('SchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: Match.stringLikeRegexp('{"DbClusterIdentifier":"db-cluster-1a"}'),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    // Start Instance Schedule testing
    it('Should have Instance Start Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-start-db-instance-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'UTC',
        ScheduleExpression: 'cron(50 7 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:rds:startDBInstance',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('SchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: Match.stringLikeRegexp('{"DbInstanceIdentifier":"db-instance-1a"}'),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    // Stop Instance Schedule testing
    it('Should have Instance Stop Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-stop-db-instance-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'UTC',
        ScheduleExpression: 'cron(10 21 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:rds:stopDBInstance',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('SchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: Match.stringLikeRegexp('{"DbInstanceIdentifier":"db-instance-1a"}'),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    it('Should match snapshot', async () => {
      expect(template.toJSON()).toMatchSnapshot('default');
    });

  });

});