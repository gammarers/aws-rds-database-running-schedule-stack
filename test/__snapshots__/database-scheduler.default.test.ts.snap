// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RdsDatabaseRunningScheduler Default Testing default schedule Should match snapshot 1`] = `
{
  "Mappings": {
    "ServiceprincipalMap": {
      "af-south-1": {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-south-2": {
        "states": "states.ap-south-2.amazonaws.com",
      },
      "ap-southeast-1": {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ap-southeast-4": {
        "states": "states.ap-southeast-4.amazonaws.com",
      },
      "ca-central-1": {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-central-2": {
        "states": "states.eu-central-2.amazonaws.com",
      },
      "eu-north-1": {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "il-central-1": {
        "states": "states.il-central-1.amazonaws.com",
      },
      "me-central-1": {
        "states": "states.me-central-1.amazonaws.com",
      },
      "me-south-1": {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-west-1": {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "SchedulerExecutionRole0AF092B7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "scheduler.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "RDS DB Running scheduler execution role",
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "states:StartExecution",
                  "Effect": "Allow",
                  "Resource": {
                    "Ref": "StateMachine2E01A3A5",
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "state-machine-exec-policy",
          },
        ],
        "RoleName": "rds-db-running-scheduler-313cb41a-exec-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "StartDBClusterSchedule": {
      "Properties": {
        "Description": "start db cluster schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "rds-db-cluster-start-313cb41a-schedule",
        "ScheduleExpression": "cron(50 7 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "UTC",
        "State": "ENABLED",
        "Target": {
          "Arn": {
            "Ref": "StateMachine2E01A3A5",
          },
          "Input": "{"Params":{"TagKey":"WorkHoursRunning","TagValues":["YES"],"Mode":"Start","Type":"Cluster"}}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "SchedulerExecutionRole0AF092B7",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "StartDBInstanceSchedule": {
      "Properties": {
        "Description": "start db instance schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "rds-db-instance-start-313cb41a-schedule",
        "ScheduleExpression": "cron(50 7 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "UTC",
        "State": "ENABLED",
        "Target": {
          "Arn": {
            "Ref": "StateMachine2E01A3A5",
          },
          "Input": "{"Params":{"TagKey":"WorkHoursRunning","TagValues":["YES"],"Mode":"Start","Type":"Instance"}}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "SchedulerExecutionRole0AF092B7",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "StateMachine2E01A3A5": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "StateMachineRoleDefaultPolicyDF1E6607",
        "StateMachineRoleB840431D",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"ModeChoice","States":{"ModeChoice":{"Type":"Choice","Choices":[{"Variable":"$.Params.Type","StringEquals":"Instance","Next":"GetDBInstanceResources"},{"Variable":"$.Params.Type","StringEquals":"Cluster","Next":"GetDBClusterResources"}],"Default":"TypeChoiceFail"},"TypeChoiceFail":{"Type":"Fail","Cause":"input Params Type is unmatch."},"GetDBInstanceResources":{"Next":"InstancesMap","Type":"Task","ResultPath":"$.Result","ResultSelector":{"TargetResources.$":"$..ResourceTagMappingList[*].ResourceARN"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:resourcegroupstaggingapi:getResources","Parameters":{"ResourceTypeFilters":["rds:db"],"TagFilters":[{"Key.$":"$.Params.TagKey","Values.$":"$.Params.TagValues"}]}},"InstancesMap":{"Type":"Map","End":true,"Parameters":{"TargetResource.$":"$$.Map.Item.Value","InputParams.$":"$.Params"},"ItemsPath":"$.Result.TargetResources","ItemProcessor":{"ProcessorConfig":{"Mode":"INLINE"},"StartAt":"GetDBInstanceIdentifier","States":{"GetDBInstanceIdentifier":{"Type":"Pass","ResultPath":"$.Result","Parameters":{"TargetDBInstanceIdentifier.$":"States.ArrayGetItem(States.StringSplit($.TargetResource, ':'), 6)"},"Next":"DescribeDBInstances"},"DescribeDBInstances":{"Next":"DBInstanceStatusChoice","Type":"Task","ResultPath":"$.Result.DBInstance","ResultSelector":{"CurrentStatus.$":"$.DbInstances[0].DbInstanceStatus"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:rds:describeDBInstances","Parameters":{"DbInstanceIdentifier.$":"$.Result.TargetDBInstanceIdentifier"}},"DBInstanceStatusChangeWait":{"Type":"Wait","Seconds":60,"Next":"DescribeDBInstances"},"StartDBInstance":{"Next":"DBInstanceStatusChangeWait","Type":"Task","ResultPath":"$.Result.StartDBInstance","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:rds:startDBInstance","Parameters":{"DbInstanceIdentifier.$":"$.Result.TargetDBInstanceIdentifier"}},"DBInstanceStatusChoice":{"Type":"Choice","Choices":[{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Start"},{"Variable":"$.Result.DBInstance.CurrentStatus","StringEquals":"stopped"}],"Next":"StartDBInstance"},{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Stop"},{"Variable":"$.Result.DBInstance.CurrentStatus","StringEquals":"available"}],"Next":"StopDBInstance"},{"Or":[{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Start"},{"Variable":"$.Result.DBInstance.CurrentStatus","StringEquals":"available"}]},{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Stop"},{"Variable":"$.Result.DBInstance.CurrentStatus","StringEquals":"stopped"}]}],"Next":"DBInstanceStateChangeSucceed"},{"Or":[{"And":[{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Start"},{"Or":[{"Variable":"$.Result.DBInstance.CurrentStatus","StringEquals":"starting"},{"Variable":"$.Result.DBInstance.CurrentStatus","StringEquals":"configuring-enhanced-monitoring"},{"Variable":"$.Result.DBInstance.CurrentStatus","StringEquals":"backing-up"},{"Variable":"$.Result.DBInstance.CurrentStatus","StringEquals":"modifying"}]}]}]},{"And":[{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Stop"},{"Or":[{"Variable":"$.Result.DBInstance.CurrentStatus","StringEquals":"modifying"},{"Variable":"$.Result.DBInstance.CurrentStatus","StringEquals":"stopping"}]}]}]}],"Next":"DBInstanceStatusChangeWait"}],"Default":"DBInstanceStatusFail"},"DBInstanceStatusFail":{"Type":"Fail","Cause":"db instance status fail."},"StopDBInstance":{"Next":"DBInstanceStatusChangeWait","Type":"Task","ResultPath":"$.Result.StopDBInstance","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:rds:stopDBInstance","Parameters":{"DbInstanceIdentifier.$":"$.Result.TargetDBInstanceIdentifier"}},"DBInstanceStateChangeSucceed":{"Type":"Succeed"}}},"MaxConcurrency":10},"GetDBClusterResources":{"Next":"ClustersMap","Type":"Task","ResultPath":"$.Result","ResultSelector":{"TargetResources.$":"$..ResourceTagMappingList[*].ResourceARN"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:resourcegroupstaggingapi:getResources","Parameters":{"ResourceTypeFilters":["rds:cluster"],"TagFilters":[{"Key.$":"$.Params.TagKey","Values.$":"$.Params.TagValues"}]}},"ClustersMap":{"Type":"Map","End":true,"Parameters":{"TargetResource.$":"$$.Map.Item.Value","InputParams.$":"$.Params"},"ItemsPath":"$.Result.TargetResources","ItemProcessor":{"ProcessorConfig":{"Mode":"INLINE"},"StartAt":"GetDBClusterIdentifier","States":{"GetDBClusterIdentifier":{"Type":"Pass","ResultPath":"$.Result","Parameters":{"TargetDBClusterIdentifier.$":"States.ArrayGetItem(States.StringSplit($.TargetResource, ':'), 6)"},"Next":"DescribeDBClusters"},"DescribeDBClusters":{"Next":"DBClusterStatusChoice","Catch":[{"ErrorEquals":["Rds.DbClusterNotFoundException"],"Next":"HandleDbClusterNotFound"}],"Type":"Task","ResultPath":"$.Result.DBCluster","ResultSelector":{"CurrentStatus.$":"$.DbClusters[0].Status"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:rds:describeDBClusters","Parameters":{"DbClusterIdentifier.$":"$.Result.TargetDBClusterIdentifier"}},"DBClusterStatusChangeWait":{"Type":"Wait","Seconds":60,"Next":"DescribeDBClusters"},"StartDBCluster":{"Next":"DBClusterStatusChangeWait","Type":"Task","ResultPath":"$.Result.StartDBCluster","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:rds:startDBCluster","Parameters":{"DbClusterIdentifier.$":"$.Result.TargetDBClusterIdentifier"}},"DBClusterStatusChoice":{"Type":"Choice","Choices":[{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Start"},{"Variable":"$.Result.DBCluster.CurrentStatus","StringEquals":"stopped"}],"Next":"StartDBCluster"},{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Stop"},{"Variable":"$.Result.DBCluster.CurrentStatus","StringEquals":"available"}],"Next":"StopDBCluster"},{"Or":[{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Start"},{"Variable":"$.Result.DBCluster.CurrentStatus","StringEquals":"available"}]},{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Stop"},{"Variable":"$.Result.DBCluster.CurrentStatus","StringEquals":"stopped"}]}],"Next":"DBClusterStateChangeSucceed"},{"Or":[{"And":[{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Start"},{"Or":[{"Variable":"$.Result.DBCluster.CurrentStatus","StringEquals":"starting"},{"Variable":"$.Result.DBCluster.CurrentStatus","StringEquals":"configuring-enhanced-monitoring"},{"Variable":"$.Result.DBCluster.CurrentStatus","StringEquals":"backing-up"},{"Variable":"$.Result.DBCluster.CurrentStatus","StringEquals":"modifying"}]}]}]},{"And":[{"And":[{"Variable":"$.InputParams.Mode","StringEquals":"Stop"},{"Or":[{"Variable":"$.Result.DBCluster.CurrentStatus","StringEquals":"modifying"},{"Variable":"$.Result.DBCluster.CurrentStatus","StringEquals":"stopping"}]}]}]}],"Next":"DBClusterStatusChangeWait"}],"Default":"DBClusterStatusFail"},"DBClusterStatusFail":{"Type":"Fail","Cause":"db clusater status fail."},"StopDBCluster":{"Next":"DBClusterStatusChangeWait","Type":"Task","ResultPath":"$.Result.StopDBCluster","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:rds:stopDBCluster","Parameters":{"DbClusterIdentifier.$":"$.Result.TargetDBClusterIdentifier"}},"DBClusterStateChangeSucceed":{"Type":"Succeed"},"HandleDbClusterNotFound":{"Type":"Pass","Result":{"message":"DB Cluster not found, but continuing..."},"End":true}}},"MaxConcurrency":10}}}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "StateMachineRoleB840431D",
            "Arn",
          ],
        },
        "StateMachineName": "rds-db-running-schedule-313cb41a-state-machine",
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "StateMachineRoleB840431D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "RDS DB Running chedule machine role.",
        "RoleName": "rds-db-running-schedule-313cb41a-state-machine-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineRoleDefaultPolicyDF1E6607": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "tag:GetResources",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "rds:describeDBInstances",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:rds:*:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":db:*",
                  ],
                ],
              },
            },
            {
              "Action": "rds:startDBInstance",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:rds:*:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":db:*",
                  ],
                ],
              },
            },
            {
              "Action": "rds:stopDBInstance",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:rds:*:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":db:*",
                  ],
                ],
              },
            },
            {
              "Action": "rds:describeDBClusters",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:rds:*:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":cluster:*",
                  ],
                ],
              },
            },
            {
              "Action": "rds:startDBCluster",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:rds:*:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":cluster:*",
                  ],
                ],
              },
            },
            {
              "Action": "rds:stopDBCluster",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:rds:*:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":cluster:*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "default-policy",
        "Roles": [
          {
            "Ref": "StateMachineRoleB840431D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StopDBClusterSchedule": {
      "Properties": {
        "Description": "stop db cluster schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "rds-db-cluster-stop-313cb41a-schedule",
        "ScheduleExpression": "cron(10 21 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "UTC",
        "State": "ENABLED",
        "Target": {
          "Arn": {
            "Ref": "StateMachine2E01A3A5",
          },
          "Input": "{"Params":{"TagKey":"WorkHoursRunning","TagValues":["YES"],"Mode":"Stop","Type":"Cluster"}}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "SchedulerExecutionRole0AF092B7",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "StopDBInstanceSchedule": {
      "Properties": {
        "Description": "stop db instance schedule.",
        "FlexibleTimeWindow": {
          "Mode": "OFF",
        },
        "Name": "rds-db-instance-stop-313cb41a-schedule",
        "ScheduleExpression": "cron(10 21 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "UTC",
        "State": "ENABLED",
        "Target": {
          "Arn": {
            "Ref": "StateMachine2E01A3A5",
          },
          "Input": "{"Params":{"TagKey":"WorkHoursRunning","TagValues":["YES"],"Mode":"Stop","Type":"Instance"}}",
          "RetryPolicy": {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": {
            "Fn::GetAtt": [
              "SchedulerExecutionRole0AF092B7",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
